Fix gathering inherited properties

IgnatBeresnev opened the pull request:
Continuation from https://github.com/Kotlin/dokka/pull/2353, rebased and with removed code duplication

tnturner83 commented:
> Continuation from https://github.com/Kotlin/dokka/pull/2353, rebased and with removed code duplication

function getWebviewContent() {
  return `<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">

    <meta http-equiv="Content-Security-Policy" content="default-src 'none';">

    <meta name="viewport" content="width=device-width, initial-scale=1.0">

    <title>Cat Coding</title>
</head>
<body>
    ...
</body>
</html>`;
}

IgnatBeresnev commented:
* Removed the transformer, moved the same logic to the translators (descriptors and psi)
* Added handling of boolean accessor names, not supported previously, which led to bugs
* Some refactoring along the way

IgnatBeresnev commented on plugins/base/src/main/kotlin/translators/psi/DefaultPsiToDocumentableTranslator.kt:
```diff
@@ -521,68 +586,80 @@ class DefaultPsiToDocumentableTranslator(
+                visibility = psi.getVisibility().toSourceSetDependent(),
+                type = getBound(psi.type),
+                receiver = null,
+                setter = setter,
+                getter = getter,
```
Before adding parameter names, these two were mixed, so `setter = getter` and `getter = setter`. Fortunately, the order was mixed in the arguments as well, so all the tests passed.

To avoid such problems in the future reviews and modifications, added parameter names to all documentable creation

IgnatBeresnev commented on plugins/base/src/main/kotlin/translators/descriptors/DescriptorAccessorConventionUtil.kt:
```diff
@@ -0,0 +1,79 @@
+    } else {
+        null
+    }
+    val kotlinPropName = propertyNameByGetMethodName(functionDescriptor.name)?.asString()
```
In the example you gave - sure, but not in this particular place and it's not an accessor by current definition for multiple reasons:
1. `val v` is immutable, should be `var`
2. `setV` is not a proper setter since it doesn't have an argument. Should be `fun setV(v: Int)`, otherwise this could be just a simple method
3. If you change change the example to account for 1 and 2, the code will not compile with declaration clash error. I believe you can't write an explicit accessor in Kotlin, `get()` and `set()` should be used, but I might be wrong
4. Last but not least, `val v` is public, there's a check down below that that accessors should not be public api.

vmishenev replied:
>  it's not an accessor

Yes, it was made intentionally. 
In my example the **regular** function  `setV()` will be lost in documentation.

IgnatBeresnev replied:
Wow, nice! I totally missed that. Added a fix + tests

IgnatBeresnev commented on plugins/base/src/test/kotlin/superFields/PsiSuperFieldsTest.kt:
```diff
@@ -0,0 +1,156 @@
+                val property = inheritorProperties.single { it.name == "isActive" }
+
+                assertNotNull(property.getter)
+                assertEquals("isActive", property.getter!!.name)
```
Indeed, fixed

IgnatBeresnev commented on plugins/base/src/main/kotlin/translators/descriptors/DescriptorAccessorConventionUtil.kt:
```diff
@@ -0,0 +1,79 @@
+    val accessors: Map<PropertyDescriptor, List<FunctionDescriptor>>
+)
+
+internal fun splitFunctionsAndAccessors(
```
Yeah, unfortunately it looks like copy-paste, but in reality it's not really - it just uses the same principles

One deals with PSIs and the other one deals with Descriptors - I didn't find any common supertypes so that I could generify the logic. Arguably, I could add a bunch of our own abstractions and try to add complex generics, but it seems overkill for such a simple and small function

If you can suggest something more specific, we can try, but I couldn't find an elegant solution

vmishenev replied:
Unfortunately, current design depends on a low level things instead of our documenable model(


Maybe a inline function with reified types could be fit here.

IgnatBeresnev commented on plugins/base/src/main/kotlin/translators/psi/DefaultPsiToDocumentableTranslator.kt:
```diff
@@ -521,68 +586,80 @@ class DefaultPsiToDocumentableTranslator(
+                extra = psi.additionalExtras().let {
+                    val psiAnnotations = psi.annotations.toList().toListOfAnnotations()
+                    val extraModifierAnnotations = it.toListOfAnnotations()
+                    val jvmFieldAnnotation = psi.annotations.toList().findJvmFieldAnnotation()
```
No idea unfortunately, this logic was written by Andrzej, but there are some tests written for JvmField and they pass, so it must be working.

If you have a more specific question / case / concern, I can check

IgnatBeresnev commented on plugins/base/src/main/kotlin/translators/psi/DefaultPsiToDocumentableTranslator.kt:
```diff
@@ -223,24 +238,60 @@ class DefaultPsiToDocumentableTranslator(
                 val documentation = javadocParser.parseDocumentation(this).toSourceSetDependent()
                 val allFunctions = async {
-                    regularFunctions.parallelMapNotNull {
+                    val mappedRegularFunctions = regularFunctions.parallelMapNotNull {
```
Sure, fixed

IgnatBeresnev commented on plugins/base/src/main/kotlin/translators/descriptors/DefaultDescriptorToDocumentableTranslator.kt:
```diff
@@ -420,14 +425,51 @@ private class DokkaDescriptorVisitor(
+
+        suspend fun getImplicitAccessorGetter() =
+            implicitAccessors
+                .firstOrNull { it.isGetterFor(originalDescriptor) }
```
Not sure what you mean unfortunately, but I've had a look at the functions it and changed `singleOrNull` to `firstIsInstanceOrNull` to avoid an intermediate operation, maybe that answers the question. If not, can you please demonstrate?

vmishenev replied:
On line 449 you use `singleOrNull`, but `singleOrNull` is used here. What function should be applied?

vmishenev commented on plugins/base/src/main/kotlin/translators/descriptors/DescriptorAccessorConventionUtil.kt:
```diff
@@ -0,0 +1,83 @@
+internal fun FunctionDescriptor.isGetterFor(property: PropertyDescriptor): Boolean {
+    return this.returnType == property.returnType
+            && this.valueParameters.isEmpty()
+            && !property.visibility.isPublicAPI
```
Do we have a test of this logic (there is a non-public property)?

IgnatBeresnev replied:
Yeah, Andrej wrote a test for that and it failed when I forgot to implement it :) 

The name is `kotlin inheriting java should not append anything since field is public` inside `DescriptorSuperPropertiesTest`
